kv: move setting the gateway node's observed timestamp to txn coord

spencerkimball opened the pull request:
This allows us to avoid setting the observed timestamp at remote leaf
transactions in a distributed SQL setting.

Release note: None

spencerkimball committed:
kv: move setting the gateway node's observed timestamp to client.Txn

This allows us to avoid setting the observed timestamp at remote leaf
transactions in a distributed SQL setting.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22981)
<!-- Reviewable:end -->

andreimatei commented:
Instead of putting this code in what looks to me like a pretty random place in TCS, can we put it in the Trasanction's ctor instead (`MakeTransaction()`)? I think that's where it belongs, plus it would only run once and not every time that "validate" function gets called (not to mention that "validate" doesn't suggest modifications).

I was commenting elsewhere that a unit test for how the transaction is initialized would be nice (you're taking away the one we had). Besides that, I think a test showing an RWU from DistSQL would be great. I know we were talking about it the other day, but now I forgot why we thought that'd be hard to write... Assuming we have any test that produces an RWU, getting such an error through DistSQL should be no harder - make a cluster with 2 nodes, create a table and use `ALTER TABLE SPLIT` to put the lease on node 1, start a txn on node 2, then write on node 1, then read through the node 2 txn and expect the error... Right?

Finally, not strictly related, but if you take the opportunity to address #15013 (the issues that we're not collecting observed timestamps through DistSQL), you get my gratitude. Or maybe has this been addressed already through whatever you've done to collect the read spans? If it hasn't, I imagine it'd be easy given that infrastructure.

---

Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender_test.go, line 472 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-L62msrWhDQPxS4Qxcr0:-L62msrWhDQPxS4Qxcr1:b-zgmfsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2900c8df91f1131461273e51faef47808571b4d6/pkg/kv/dist_sender_test.go#L472)):*
> ```Go
> )
> 
> func TestOwnNodeCertain(t *testing.T) {
> ```

can't this test be adapted to go through the TCS? Or if you take the suggestion of moving the code to `MakeTransaction()`, we can have a tiny unit test for that. 

---

*[pkg/kv/txn_coord_sender.go, line 225 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-L62mBnJAATg6nWPdsSN:-L62mBnJAATg6nWPdsSO:bqlejln) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fbf06fd25f237e888f8f2a478fee03dd78e3922/pkg/kv/txn_coord_sender.go#L225)):*
> ```Go
> 	wrapped           client.Sender
> 	clock             *hlc.Clock
> 	gossip            *gossip.Gossip
> ```

if this can be nil, please add a comment about what that means.
But I don't understand why this factory and the TCS need gossip - they just need the node id, no?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-:-L62p0iAarJ_5jHnkVuJ:b-xi8rka)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
cc @tschottdorf so he knows his complaining is not being ignored :). Also, he should take offense that this PR claims to "fix" his PR #21266 :)

---

Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-:-L62sANN-Beft6f_MmpF:b-2a0pcd)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Why wouldn't this fix #21266? Mind you I haven't tested that it does, but from my reading of that issue, this certainly seems like the fix.

#15013 is fixed (I closed that issue).

Good points in your comments. I've rewritten this change to set the observed timestamp optimization in `client.NewTxn`.

---

Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/kv/dist_sender_test.go, line 472 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-L62msrWhDQPxS4Qxcr0:-L67re08VkVUiQWkmY86:brii3cv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2900c8df91f1131461273e51faef47808571b4d6/pkg/kv/dist_sender_test.go#L472)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can't this test be adapted to go through the TCS? Or if you take the suggestion of moving the code to `MakeTransaction()`, we can have a tiny unit test for that. 
</blockquote></details>

This test was garbage. I've added a basic test in `internal/client` for this functionality.

---

*[pkg/kv/txn_coord_sender.go, line 225 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-L62mBnJAATg6nWPdsSN:-L67rjjOXvMF-2mgUL2a:b-fk1166) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fbf06fd25f237e888f8f2a478fee03dd78e3922/pkg/kv/txn_coord_sender.go#L225)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if this can be nil, please add a comment about what that means.
But I don't understand why this factory and the TCS need gossip - they just need the node id, no?
</blockquote></details>

no longer applicable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-:-L67_4_dtMkvCGJlATE-:b-7lo8o6)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

> Why wouldn't this fix #21266? 

I was just saying that that's technically a PR, not an issue :)

I still think a DistSQL test would be nice, unless it proves to be hard to write for some reason.

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-:-L689svEhx8l-TTwkeyj:b-gqfj7u)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
I agree a dist SQL test should be written, ideally one which is able to capture the same dynamic as in #21266. Care to create an actual issue for that?

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-:-L68Aj8qKOl26kQr7BKM:br381pa)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@andreimatei actually if you can point me at a dist SQL test which already does something vaguely similar, I can make it happen.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 11 files at r1, 5 of 13 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-:-L6DUzZgJVQT96gESL2N:bg7igkv)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Spencer I noticed a random thing that perhaps asks for improvement - we have the `BatchRequest.Header.GatewayNodeID` field, serving (I think) for collecting some origin statistics. Do you reckon there's an opportunity to consolidate that with the `Transaction`'s newly discovered interest in knowing what the gateway is? The `GatewayNodeID` field is per batch, and not all batches are transactional, but I think we don't care about maintaining those stats on non-transactional requests.

For testing, you can check out `TestClusterFlow` for a relatively lower level test that builds flows by hand and runs them on different nodes in a cluster - and so you can mimic that to see that a `TableReader`returns the right error. Or you can check out `TestDistSQLRangeCachesIntegrationTest` for a higher-level test that controls data distribution and then performs a DistSQL query over manually-placed data.
Does that help?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22981#-:-L6DuRduNnUnhiKzUe8E:b-ic1qyn)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@andreimatei we're not actually storing the node ID explicitly, and we're not specifying anywhere that the one certain timestamp is actually the gateway node. I think the current state of things is too unformed to conflate these two purposes, despite there clearly being some overlap.

I'll work on a distSQL unittest in another PR.
